packages/evm/src/eof/container.ts

```diff
@@ -0,0 +1,229 @@
+import { KIND_CODE, KIND_DATA, KIND_TYPE, MAGIC, TERMINATOR, VERSION } from './constants'
+import { EOFError, validationError } from './errors'
+
+class StreamReader {
+  private data: Buffer
+  private ptr: number
+  constructor(stream: Buffer) {
+    this.data = stream
+    this.ptr = 0
+  }
+
+  readBytes(amount: number, errorStr?: string) {
+    const end = this.ptr + amount
+    if (end > this.data.length) {
+      validationError(EOFError.OutOfBounds, this.ptr, errorStr)
+    }
+    const ptr = this.ptr
+    this.ptr += amount
```

Comment:
Is this correct that the pointer is always increased *before* the read? 🤔 Same for `readUint()` and `readUint16()`.

Reply:
(but maybe (likely) it does since when I change it I got errors e.g. on the version number. Just looks counter-intuitive to me)

Reply:
Yeah, after some more code reading: I find this mixing of verifying and shifting the pointer unlucky. A method should *either* verify some bytes or a single bytes and not touch the state (shift the pointer) or read bytes (from the position the pointer is on and *at the end* move the pointer.

Everything else is getting error prone pretty quickly.

Let me know if I am missing something basic here.

Reply:
Update: ah, ok, I misread the code, didn't get the `ptr` vs `this.prt` separation and was confused by this "first increasing the pointer" code order. I would nevertheless think this would be worth to rewrite a bit to e.g.:

```typescript
const data = this.data.readUint16BE(this.ptr)
this.ptr += 2
return data
```

This otherwise leads to confusion.